### ISSUE TITLE
Orphaned section headers

### ISSUE THREAD

**Issue by hudcap** [moderncv/moderncv#10]
Pagebreaks can happen immediately after a (sub)section header.
Reported as an issue on the original project [here](https://github.com/xdanaux/moderncv/issues/5) and [here](https://github.com/xdanaux/moderncv/issues/82), and on [StackOverflow](https://tex.stackexchange.com/questions/342971/class-moderncv-orphan-headings).

Not sure if this is helpful, or if you plan on "importing" all the issues from the original project (if that's even possible).
I opened this because every time I change my CV, it messes up the pagebreaks again, but feel free to close it if this isn't helpful.

Thanks for reviving this project!!

**Comment by cryptointerest** [comment]
noted, but I have no idea how to fix this. I resort to manually adjusting things.

**Comment by hudcap** [comment]
This hack, based on https://github.com/xdanaux/moderncv/issues/82#issuecomment-570812909, "solved" my issue:

```
\usepackage{needspace}
\let\oldsection=\section
\renewcommand{\section}{%
  \needspace{3\baselineskip}
  \oldsection
}
```

**Comment by cryptointerest** [comment]
is this in all styles or one specific style?

**Comment by cryptointerest** [comment]
ok, seems like a proper solution is contained in [this post](https://tex.stackexchange.com/questions/348994/understanding-needspace). Posting this for reference.

**Comment by hudcap** [comment]
> is this in all styles or one specific style?
> I mean, I know that page breaks are horrible with the fancy style, but this is due to a dirty hack that as of now has no other solution.
> So in fancy, manual adjustments are needed for sure.

I've only ever used the classic style, so I can only confirm the problem and the solution for classic.

**Comment by cryptointerest** [comment]
I think I fixed it, at least it works with my own CV. 
Could someone test if it works with their CV, too? The new code can be found [here](https://github.com/cryptointerest/moderncv/tree/orphaned_section_fix).

**Comment by hudcap** [comment]
I just tested it.
With the old code, I had an orphaned section and subsection header. The new code "unorphaned" the subsection header and placed it on the next page, but left the section header orphaned on the previous page.
I don't fully understand how this works, but my guess is that there was "enough" penalty for one header, but not enough for two.

Hope this helps.
I can try to create a MWE if you can't replicate this.

Thanks!

**Comment by cryptointerest** [comment]
uh interesting. I don't use subsections in my cv. maybe I should. 
could you try using the command 
```latex
\ setlength {\ cvstretchability }{<length>}
```
in your document where <length> can be e.g. `\ baselineskip ` or another length like `100pt`

**Comment by hudcap** [comment]
I had to go all the way up to `\setlength{\cvstretchability}{5\baselineskip}` to get it to work, but then another section that previosuly displayed 11 lines of text on page 2 was completely pushed to page 3, leaving a lot of empty space on page 2.

**Comment by cryptointerest** [comment]
mhh ok, so I introduced separate stretchabilities for both sections and subsections and played around with that, but the result highly depends on the content of the CV. Notably, I don't see any generic MWE. 
Example: 
In my section IT Skills I use my new skillmatrix feature. Due to the legend being quite a big block I get an orphaned section. 
But if I added another \cventry{}{} before the skill matrix (or instead of that), that item would still fit. Therefore latex does not break the page and rightly so. If I added instead of \cventry a \subsection command, again, we would end up with an orphaned section, subsection pair. 
This leads to the following problem: if I increased the  stretchability for \section (and possibly \subsection) by default, I'd force a page break _by default_ in all these situations. In the first and third setting this might be desired, but in the second setting wastes too much space. 

So in short, I don't see a generic _optimal_ for all cases. Notably, the „orphaned section“ problem does not seem to have a solution. Therefore, I believe it is best to provide the user with the options
```latex
\setlength{\cvsectionstretchability}{<some length>}
\setlength{\cvsubsectionstretchability}{<some other length>}
```
 to adjust the stretchability when needed, but not set it by default. This will be explained in the new user manual. 
 This means a certain element of _manual adjustment_ for each CV similar to including a manual page breaks or adjusting `\arraystretch` but I don't see another solution.

**Comment by hudcap** [comment]
Thanks for all your work on this!

Okay, I see what you're saying.

I don't fully understand this topic, so please forgive me if I'm using the wrong terminology, but shouldn't there be a way of imposing an infinite penalty against a page break immediately following a (sub)section?

**Comment by cryptointerest** [comment]
There isn't directly. We can only tell latex to break `\section` and `\subsection` to the new page if latex does not have „enough“ space left on the current page for an entire line (or possibly several lines. This is defined by the stretchability command).

I updated the code to 
```latex
\setlength{\cvsectionstretchability}{.9\baselineskip}
\setlength{\cvsubsectionstretchability}{.9\baselineskip}
 ```
by default. This should solve orphaned `\section` and `\subsection` commands that are used _alone_ for _most people_. (Feel free to test after [downloading the updated code](https://github.com/cryptointerest/moderncv/tree/update_documentation)  by commenting out your subsection command following the section command in your CV).

However latex does not check for content. So in case of using
 ```latex
\section{blub}
\subsection{blubblub}
 ```
directly one after the other (like you do in your CV), the subsection will be unorphaned leaving _enough space_ on the previous page for latex _not to also_ break `\section{blub}` to the new page. One way of solving this is to increase `\cvsectionstretchability` to an artificially high value to force the break of the section header. (try this!).
However, this will change the behaviour for _all_ `\section` commands in your CV and might cause a page break in a situation where it would not have been necessary (down the line on another page), e.g. in a case like 
 ```latex
\section{blub}
\cvitem{blubblub}
 ```
in which the item  previously would have fit on the page. You described this effect yourself in one [of your previous posts](https://github.com/moderncv/moderncv/issues/10#issuecomment-768397999). 
I therefore believe that in that special case (your case) it is better to solve the page break by a `\newpage` command.
